crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/migrations.rs

```diff
@@ -0,0 +1,148 @@
+use std::fs;
+use std::path::Path;
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use tracing::{debug, error};
+
+use super::LedgerDB;
+use crate::ledger_db::SharedLedgerOps;
+use crate::rocks_db_config::RocksdbConfig;
+
+/// Alias for migration name type
+pub type MigrationName = String;
+/// Alias for migration version type
+pub type MigrationVersion = u64;
+
+/// A trait that should be implemented by migrations.
+pub trait LedgerMigration {
+    /// Provide an identifier for this migration
+    fn identifier(&self) -> (MigrationName, MigrationVersion);
+    /// Execute current migration on ledger DB
+    fn execute(&self, ledger_db: Arc<LedgerDB>) -> anyhow::Result<()>;
+}
+
+/// Handler for ledger DB migrations.
+///
+/// This implements migrations in an atomic fashion.
+/// Meaning that, if any migration would fail, the whole process
+/// is rolled back to the previous version, rendering the changes
+/// made by any run migration useless.
+pub struct LedgerDBMigrator<'a> {
+    ledger_path: &'a Path,
+    migrations: &'static Vec<Box<dyn LedgerMigration + Send + Sync + 'static>>,
+}
+
+impl<'a> LedgerDBMigrator<'a> {
+    /// Create new instance of migrator
+    pub fn new(
+        ledger_path: &'a Path,
+        migrations: &'static Vec<Box<dyn LedgerMigration + Send + Sync + 'static>>,
+    ) -> Self {
+        Self {
+            ledger_path,
+            migrations,
+        }
+    }
+
+    /// Run migrations
+    pub fn migrate(&self, max_open_files: Option<i32>) -> anyhow::Result<()> {
+        if self.migrations.is_empty() {
+            return Ok(());
+        }
+
+        debug!("Starting LedgerDB migrations...");
+
+        let original_path = &self.ledger_path;
+
+        let ledger_db =
+            LedgerDB::with_config(&RocksdbConfig::new(self.ledger_path, max_open_files))?;
+        let executed_migrations = ledger_db.get_executed_migrations()?;
+        // Drop the lock file
+        drop(ledger_db);
+
+        // Copy files over, if temp_db_path falls out of scope, the directory is removed.
+        let temp_db_path = tempfile::tempdir()?;
+        copy_db_dir_recursive(original_path, temp_db_path.path())?;
+
+        let new_ledger_db = Arc::new(LedgerDB::with_config(&RocksdbConfig::new(
+            temp_db_path.path(),
+            max_open_files,
+        ))?);
+
+        for migration in self.migrations {
+            if !executed_migrations.contains(&migration.identifier()) {
```

Comment:
i think erroring out on failed migration is reasonable. bc if the migration couldn't be executed successfully, then the node will fail once it needs to use one of the tables in those migrations.

Reply:
also before erroring out, it can revert to the old db

Reply:
I am not sure i understand the change request here. But in general, those lines are running migrations on the temporary directory database. If any of those migrations fail, we don't have to roll back anything we just discard the temporary directory DB updates. This is why i copy the temporary directory into the ledger DB directory as the last step.

Reply:
I mean on `let Err(e) = migration.execute(new_ledger_db.clone())` case we should return err and terminate the node imo

Reply:
Done